add basic condition-operator functions

xiam opened the pull request:
This PR enhaces `pgkit.Cond` with the ability of accepting `interface{}` on both sides. The right side might contain an operator, which, if present, will be used to compile a condition.

xiam committed:
add basic condition functions

xiam committed:
formatting

VojtechVitek reviewed (approved):
LGTM. Looks great! :) 🎉 

Can we also get an example with `.And{}` and `.Or{}`?

The idea is that the pgkit users could write all DB queries without having to squirrel directly. I'd like to be able to write queries as such (similar to upper/db):

```go
cond := db.Or{
	db.Cond{
		"friends_count": db.Gte(100),
		"deleted_at": db.IsNull(),
	},
	db.Cond{
		"role": db.In(roles.Admin, roles.SuperAdmin),
		"deleted_at": db.IsNull(),
	},	
}

users, err := data.DB.Users.FindAll(cond)
// ...
```

which reminds me -- would you guys agree to move this under `pgkit/db` package, so we can use "db" for short?

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+// NotLike represents a NOT LIKE comparison.
+func NotLike(v interface{}) squirrel.Sqlizer {
+	return &condExpr{"NOT LIKE", v}
+}
```
This is pure OCD, just an order of functions
```suggestion
// Eq represents an equality comparison.
func Eq(v interface{}) squirrel.Sqlizer {
	return &condExpr{"=", v}
}

// NotEq represents a not-equal comparison.
func NotEq(v interface{}) squirrel.Sqlizer {
	return &condExpr{"<>", v}
}

// Gt represents a greater-than comparison.
func Gt(v interface{}) squirrel.Sqlizer {
	return &condExpr{">", v}
}

// Gte represents a greater-than-or-equal comparison.
func Gte(v interface{}) squirrel.Sqlizer {
	return &condExpr{">=", v}
}

// Lt represents a less-than comparison.
func Lt(v interface{}) squirrel.Sqlizer {
	return &condExpr{"<", v}
}

// Lte represents a less-than-or-equal comparison.
func Lte(v interface{}) squirrel.Sqlizer {
	return &condExpr{"<=", v}
}

// IsNull represents an IS NULL comparison.
func IsNull() squirrel.Sqlizer {
	return &condExpr{"IS NULL", nil}
}

// IsNotNull represents an IS NOT NULL comparison.
func IsNotNull() squirrel.Sqlizer {
	return &condExpr{"IS NOT NULL", nil}
}

// Like represents a LIKE comparison.
func Like(v interface{}) squirrel.Sqlizer {
	return &condExpr{"LIKE", v}
}

// NotLike represents a NOT LIKE comparison.
func NotLike(v interface{}) squirrel.Sqlizer {
	return &condExpr{"NOT LIKE", v}
}

// ILike represents a ILIKE comparison.
func ILike(v interface{}) squirrel.Sqlizer {
	return &condExpr{"ILIKE", v}
}

// NotILike represents a NOT ILIKE comparison.
func NotILike(v interface{}) squirrel.Sqlizer {
	return &condExpr{"NOT ILIKE", v}
}
```

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+
+		if err != nil {
+			return "", nil, fmt.Errorf("error compiling node %d: %v", i, err)
+		}
```
Removed line and `%v` -> `%w`
```suggestion
		qn, argsn, err := node.ToSql()
		if err != nil {
			return "", nil, fmt.Errorf("error compiling node %d: %w", i, err)
		}
```

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+	})
+}
+
+// AnyOf represents an IN operator. The value must be a slice.
+func AnyOf(v interface{}) squirrel.Sqlizer {
+	return sqlExprFn(func() (string, []interface{}, error) {
+		rv := reflect.ValueOf(v)
+
+		if rv.Kind() != reflect.Slice {
+			return "", nil, fmt.Errorf("value must be a slice")
+		}
```
Just a thought, maybe this could also take array.
```suggestion
// AnyOf represents an IN operator. The value must be a slice or an array.
func AnyOf(v interface{}) squirrel.Sqlizer {
	return sqlExprFn(func() (string, []interface{}, error) {
		rv := reflect.ValueOf(v)

		rk := rv.Kind()
		if rk != reflect.Slice && rk != reflect.Array {
			return "", nil, fmt.Errorf("value must be a slice or an array")
		}
```

klaidliadon replied:
what about `AnyOf[T](v []T)`?

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+	})
+}
+
+// NotAnyOf represents a NOT IN operator. The value must be a slice.
+func NotAnyOf(v interface{}) squirrel.Sqlizer {
+	return sqlExprFn(func() (string, []interface{}, error) {
+		rv := reflect.ValueOf(v)
+
+		if rv.Kind() != reflect.Slice {
+			return "", nil, fmt.Errorf("value must be a slice")
+		}
```
Just a thought, maybe this could also take array.
```suggestion
// NotAnyOf represents a NOT IN operator. The value must be a slice or an array.
func NotAnyOf(v interface{}) squirrel.Sqlizer {
	return sqlExprFn(func() (string, []interface{}, error) {
		rv := reflect.ValueOf(v)

		rk := rv.Kind()
		if rk != reflect.Slice && rk != reflect.Array {
			return "", nil, fmt.Errorf("value must be a slice or an array")
		}
```

klaidliadon replied:
same here `NotAnyOf[T](v []T)`

david-littlefarmer replied:
IMHO This wouldn't work. You can't wrap it like this.
It would want the `[]interface{}` and it wouldn't accept `[]int`, `[]string`, etc.
`[]int != []interface{}`.
Sadly we don't have constraints for default non-primitve types yet (itf they would be ever created) only for primitives :/
https://pkg.go.dev/golang.org/x/exp@v0.0.0-20240325151524-a685a6edb6d8/constraints

xiam replied:
I think this is a great idea, we can remove `AnyOf` and replace it with a generic `In[T]` and remove reflection stuff. Doesn't look bad I think. The cost is creating a copy of the arguments, which is also necessary when using `AnyOf`, so I think it's acceptable. Please take a look at the most recent examples of `In` and `NotIn`.

VojtechVitek replied:
This is lovely :)

```
cond := pgkit.Cond{"list": pgkit.In[string]("Czech Republic", "Slovakia")}
```

which can be also written as
```
cond := pgkit.Cond{"list": pgkit.In("Czech Republic", "Slovakia")}
```

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+
+	ql, argsl, err := compileLeaf(n.left)
+	if err != nil {
+		return "", nil, fmt.Errorf("error compiling left side: %v", err)
```
```suggestion
		return "", nil, fmt.Errorf("error compiling left side: %w", err)
```

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+
+	rl, argsr, err := compileLeaf(n.right)
+	if err != nil {
+		return "", nil, fmt.Errorf("error compiling right side: %v", err)
```
```suggestion
		return "", nil, fmt.Errorf("error compiling right side: %w", err)
```

david-littlefarmer commented on cond.go:
```diff
@@ -0,0 +1,253 @@
+		qn, argsn, err := node.ToSql()
+
+		if err != nil {
+			return "", nil, fmt.Errorf("error compiling node %d: %v", i, err)
```
```suggestion
			return "", nil, fmt.Errorf("error compiling node %d: %w", i, err)
```

david-littlefarmer reviewed (approved):
This is great!!! :+1:

Here are my few comments/thoughts:
3x `%v` -> `%w` in error wrappings
2x OCD nitpicks
2x thoughts, to let `AnyOf()` and `NotAnyOf()` accept also arrays - example: https://goplay.tools/snippet/AuV1NR_sQlD

klaidliadon commented:
This looks great!
I would assume this is the kind of package you import with . so that your code looks like:
```go
cond := Or{
	Cond{
		"friends_count": Gte(100),
		"deleted_at": IsNull(),
	},
	Cond{
		"role": In(roles.Admin, roles.SuperAdmin),
		"deleted_at": IsNull(),
	},	
}
```

I think it would be cool to have some version of raw that can wrap a sqlizer:
```go
q := squirrel.Select("id").From("users").Where(sq.Eq{"id":Gt(10)})

cond := Cond{id: In(Rawf("any(%s)", q))}

xiam committed:
use proper format for errors

Co-authored-by: David Sedláček <david.littlefarmer@gmail.com>

xiam committed:
change order of functions

Co-authored-by: David Sedláček <david.littlefarmer@gmail.com>

xiam committed:
replace AnyOf with generic In

xiam commented:
@klaidliadon 

> I think it would be cool to have some version of raw that can wrap a sqlizer:

Good point, will demonstrate that case.

xiam committed:
move under "db"

xiam commented:
@klaidliadon @VojtechVitek @david-littlefarmer I implemented most of your suggestions and added a few examples around conditions with subqueries, and uses of generic `In[T]`. I still have to think about simplifying some code things, but in terms of the API we should be close to a final definition, please take another look.

VojtechVitek reviewed (approved):
This is great! <3

LGTM 🎉
